spanner/spanner_arrays/main.go

```diff
@@ -0,0 +1,180 @@
+			return
+		}
+		if err != nil {
+			log.Fatalf("failed to read results: %v", err)
+		}
+
+		var country Country
+		err = row.ToStruct(&country)
+		if err != nil {
+			log.Fatalf("failed to read row into Country struct: %s", err)
+		}
+
+		var cities []string
+		for _, c := range country.Cities {
+			cities = append(cities, c.String())
+		}
+
+		log.Printf("%s: %s", country.Name, strings.Join(cities, ", "))
+	}
+}
+
+func loadPresets(ctx context.Context, db *spanner.Client) error {
+	mx := []*spanner.Mutation{
+		spanner.InsertMap("Countries", map[string]interface{}{
+			"CountryId": 49,
+			"Name":      "Germany",
+		}),
+		spanner.InsertMap("Cities", map[string]interface{}{
+			"CountryId": 49,
+			"CityId":    100,
+			"Name":      "Berlin",
+		}),
+		spanner.InsertMap("Cities", map[string]interface{}{
+			"CountryId": 49,
+			"CityId":    101,
+			"Name":      "Hamburg",
+		}),
+		spanner.InsertMap("Cities", map[string]interface{}{
+			"CountryId": 49,
+			"CityId":    102,
+			"Name":      "Dresden",
+		}),
+		spanner.InsertMap("Countries", map[string]interface{}{
+			"CountryId": 44,
+			"Name":      "United Kingdom",
+		}),
+		spanner.InsertMap("Cities", map[string]interface{}{
+			"CountryId": 44,
+			"CityId":    200,
+			"Name":      "London",
+		}),
+		spanner.InsertMap("Cities", map[string]interface{}{
+			"CountryId": 44,
+			"CityId":    201,
+			"Name":      "Liverpool",
+		}),
+		spanner.InsertMap("Cities", map[string]interface{}{
+			"CountryId": 44,
+			"CityId":    202,
+			"Name":      "Bristol",
+		}),
+		spanner.InsertMap("Cities", map[string]interface{}{
+			"CountryId": 44,
+			"CityId":    203,
+			"Name":      "Newcastle",
+		}),
+	}
+
+	_, err := db.Apply(ctx, mx)
+	return err
+}
+
+func createDatabase(ctx context.Context, adminClient *database.DatabaseAdminClient, db string) error {
+	matches := regexp.MustCompile("^(.*)/databases/(.*)$").FindStringSubmatch(db)
+	if matches == nil || len(matches) != 3 {
+		log.Fatalf("Invalid database id %s", db)
+	}
+
+	op, err := adminClient.CreateDatabase(ctx, &adminpb.CreateDatabaseRequest{
+		Parent:          matches[1],
+		CreateStatement: fmt.Sprintf("CREATE DATABASE `%s`", matches[2]),
+		ExtraStatements: []string{
+			`CREATE TABLE Countries (
+				CountryId 	INT64 NOT NULL,
+				Name   		STRING(1024) NOT NULL
+			) PRIMARY KEY (CountryId)`,
+			`CREATE TABLE Cities (
+				CountryId	INT64 NOT NULL,
+				CityId		INT64 NOT NULL,
+				Name			STRING(MAX),
+			) PRIMARY KEY (CountryId, CityId),
+			INTERLEAVE IN PARENT Countries ON DELETE CASCADE`,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	if _, err := op.Wait(ctx); err == nil {
+		log.Printf("Created database [%s]", db)
+	}
+	return err
+}
+
+func removeDatabase(ctx context.Context, adminClient *database.DatabaseAdminClient, db string) error {
+	err := adminClient.DropDatabase(ctx, &adminpb.DropDatabaseRequest{Database: db})
+	if err != nil {
+		log.Fatalf("Failed to remove database [%s]: %v", db, err)
```

Comment:
log.Fatal is fairly odd here, I'd really expect this code to return the error and the caller to handle the error (by calling log.Fatal)

Reply:
Yeah; because this is just a named defer, returning an error is relatively useless; now handling it locally without returning error.  I'm still not totally sanguine about it philosophically; maybe it would be best to just handle it as an inline function in the defer.